### PR TITLE
Efficent Backend Design Implemented

### PR DESCRIPTION
**Pull request by AaronWB**
Changed data structure from multiple hash maps into a single hash map to save on RAM.

Gave constructors for Classes, Methods, Fields where a Class Object now holds all the Data required (Methods & Parameters, Fields).

Improved functionality by adding a Field Type to Fields.

Changed data types to mainly strings to remove possibility of issues by limiting the data type.

### CODE REVIEW

File: UML editor/src/main/java/codecain/UMLMethodInfo.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+     * @param parameters the list of parameters for the method (can be null)
+     */
+    public UMLMethodInfo(String methodName, List<String> parameters) {
+        this.methodName = methodName;
+        if (parameters != null) {
+            this.parameters = new ArrayList<>(parameters);
+        } else {
+            this.parameters = new ArrayList<>();
+        }
+    }
```
Comment (mmoua10) [inline UML editor/src/main/java/codecain/UMLMethodInfo.java:35]: The parameters will never be null because the command wont work if parameters are left empty.

**Review by mmoua10 — CHANGES_REQUESTED** [review]
It seems like you just need to change the null exception.

File: UML editor/src/main/java/codecain/UMLMethodInfo.java (hunk 1 above)
Comment (AaronWB) [reply UML editor/src/main/java/codecain/UMLMethodInfo.java:35]: Thought process was I had two constructors initially that handled if we wanted to just to create the name, or name and parameters, but this simplifies it so we could use the same constructor but put null in the parameter for example if we wanted to just initialize a method, or if a method has no parameters we put null instead of inserting a string that is blank.

**Review by mmoua10 — APPROVED** [review]
Looks good, logic looks good.

**Comment by colbycrutcher** [comment]
This is way better than having multiple data structures working in conjunction. Logic all seems good from what I am seeing. Will greatly benefit the design of our code going forward.

**Comment by colbycrutcher** [comment]
UMLMethodInfo is a good way to store parameters, which we were struggling with in sprint 1.

File: UML editor/src/main/java/codecain/UMLClass.java [hunk 2]
```diff
@@ -98,28 +72,10 @@ public static void renameClass(Object oldClassName, Object newClassName) {
...
             System.out.println("Class " + newClassName + " already exists");
         } else {
-            UMLClass classObj = classMap.remove(oldClassName);
-            classObj.className = newClassName;
-            classMap.put(newClassName, classObj);
-            Map<Object, List<Object>> methods = UMLMethods.classMethods.remove(oldClassName);
-            UMLMethods.classMethods.put(newClassName, methods);
-            Map<Object, Object> fields = UMLFields.classFields.remove(oldClassName);
-            UMLFields.classFields.put(newClassName, fields);
+            UMLClassInfo classInfo = classMap.remove(oldClassName);
```
Comment (colbycrutcher) [inline UML editor/src/main/java/codecain/UMLClass.java:75]: Efficient way to rename classes, a functionality we did not have in sprint 1

File: UML editor/src/main/java/codecain/UMLClassInfo.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
+package codecain;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * The UMLClassInfo class represents the details of a UML class, including its name, fields, and methods.
+ * It provides methods to retrieve and manage the fields and methods of the class.
+ */
+public class UMLClassInfo {
```
Comment (colbycrutcher) [inline UML editor/src/main/java/codecain/UMLClassInfo.java:10]: Nice addition to our code. We had a hard time with the fields, and methods in sprint one, especially displaying them.

File: UML editor/src/main/java/codecain/UMLFieldInfo.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
+package codecain;
+
+/**
+ * The UMLFieldsInfo class represents a field in a UML class.
+ * It includes the field name and the field type.
+ */
+public class UMLFieldInfo {
```
Comment (colbycrutcher) [inline UML editor/src/main/java/codecain/UMLFieldInfo.java:7]: UMLFieldInfo is a nice addition we didn't have in sprint1. Having multiple maps last sprint had me confused about what exactly was going on, and having this clarity will nice for everyone

File: UML editor/src/test/java/UMLClassTest.java [hunk 5]
```diff
@@ -18,157 +20,180 @@ public class UMLClassTest {
...
+    public void testRemoveClass() {
+        UMLClass.addClass("TestClass1");
+        UMLClass.removeClass("TestClass1");
+        assertFalse(UMLClass.classMap.containsKey("TestClass1"), "Class should be removed from the map");
+    }
+
+    @Test
+    public void testRenameClass() {
+        UMLClass.addClass("TestClass1");
+        UMLClass.renameClass("TestClass1", "RenamedClass");
```
Comment (colbycrutcher) [inline UML editor/src/test/java/UMLClassTest.java:181]: Clear and descriptive tests, and good use of assertions. We also get good testing on edge cases

File: UML editor/src/main/java/codecain/UMLMethods.java [hunk 6]
```diff
@@ -1,225 +1,170 @@
...
-        if (isClassNameOrMethodNameOrParameterInvalid(className, methodName, parameter)) return;
-        if (!doesMethodExist(className, methodName)) return;
-        List<Object> parameters = classMethods.get(className).get(methodName);
-        if (!parameters.contains(parameter)) {
-            System.out.println("Action Canceled: Parameter " + parameter + " does not exist in method " + methodName);
-        } else {
-            parameters.remove(parameter);
+    public void removeParameter(String className, String methodName, String parameter) {
+        UMLClassInfo classInfo = getClassInfo(className);
+        if (classInfo == null) return;
```
Comment (colbycrutcher) [inline UML editor/src/main/java/codecain/UMLMethods.java:104]: Improved, using String for clarity. Simplified Logic,  and better error messages. I also like the feedback it adds to the user

**Review by colbycrutcher — COMMENTED** [review]
Overall good, cleans up our backend drastically. Looks good.
